### ISSUE TITLE
Custom "artificial cursor" option

### ISSUE THREAD

**Issue by chjj** [slap-editor/slap#138]
I thought this might be a cool option for a text editor: https://github.com/chjj/blessed/commit/d3e566b9c1672ddb9d41df25131d1710716dd3b4

Blessed now offers a custom or "artificial" cursor. It overrides the default terminal cursor. I don't think this has ever been done in any other terminal application, but I'm sure users may like it.

**Comment by kookerus** [comment]
I don't think this would really be necessary. Most terminals support changing the cursor as it is, so this really wouldn't be needed.

**Comment by chjj** [comment]
kookerus, that's actually not true. There's a reason I implemented this.

xterm only had 2 different cursors up until recently (now it supports the vertical bar), and rxvt-* does not support any OSC codes to change the cursor or to reset the cursor properly. iTerm2 can change cursors, but it only supports 2 cursors. Every terminal is different in this regard. Proprietary cursor sequences are very unreliable. The artificial cursors would allow us to create _any_ kind of cursor. The cursor could just be a boldened cell for example.

**Comment by chjj** [comment]
@kookerus, I also want to add, it costs almost nothing to implement artificial cursors. Given the amount of benefits, there's no reason not to include them.

**Comment by taneltm** [comment]
I am hoping for multi-cursor support in slap, in the future - custom cursor implementation is needed for multi-cursor anyway.

I see no reason why slap shouldn't start using custom cursors now.
We may discover bugs and could switch back to native cursor until the bugs are fixed.
If we wait until we have multi-cursor support and find a bug, it may be more trouble.

**Comment by secrettriangle** [comment]
Hmm, looks like the artificial cursor option only allows for one artificial cursor. @chjj, what do you think about extending your implementation to allow for many cursors?

**Comment by kookerus** [comment]
@chjj That makes a lot of sense. I've grown far too used to the comforts of the gnome terminal and yakuake.

**Comment by chjj** [comment]
@secrettriangle, that would still probably require a bit of revision to the current code. Turning `screen.cursor` into an array and iterating through them during rendering. That's easy enough. The tough part will be having a concept of a primary cursor. Text elements and even the core sequences like CUP assume there is on cursor, so there would need to be an extra parameter to specify which cursor needs to be moved. I can't think of a pretty way to do it right now.

I've never used sublime, but from learning what I could, it looks like the multiple cursors have a lot to do with selections rather than cursors. Which reminds me: @secrettriangle, I'm really impressed you made a text editor out of blessed since it only has very primitive features for text elements. If you want, we could start making text elements a little more advanced in core (cursor movement back and forth for example).

**Comment by taneltm** [comment]
In sublime, multiple cursors are not used only/mainly for selections. For example you can ctrl+click to spawn a new cursor (there are lots of other ways to spawn cursors), then you can move all cursors via: arrow keys, home, end, pgUp, pgDown, delete, backspace and typing characters. So it is not just for selecting.

While you see multiple cursors, there is no difference between the original and other cursors, but when you hit ESC, the other cursors disappear and the original remains.
I think we could safely assume that the first cursor in the cursor stack would be the original.

**Comment by taneltm** [comment]
One thing to keep in mind is that if 2 cursors happen to be in the same place, then they are merged.
